support/BRFileService.c

```diff
@@ -127,34 +228,158 @@ fileServiceCreate (const char *basePath,
     if (NULL == dir) return NULL;
     closedir(dir);
 
-    // Create the directory hierarchy
-    size_t pathToTypeSize = strlen(basePath) + 1 + strlen(currency) + 1 + strlen(network) + 1;
-    char dirPath[pathToTypeSize];
-
-    sprintf (dirPath, "%s/%s", basePath, currency);
-    if (-1 == directoryMake(dirPath)) return NULL;
-
-    sprintf(dirPath, "%s/%s/%s", basePath, currency, network);
-    if (-1 == directoryMake(dirPath)) return NULL;
+    // Require SQLite to support 'MULTI_THREADED' or 'SERIALIZED'.  We'll lock our connection.
+    // and thus 'MULTI_THREADED' is appropriate.
+    if (0 == sqlite3_threadsafe()) return NULL;
 
+    // Create the file service itself
     BRFileService fs = calloc (1, sizeof (struct BRFileServiceRecord));
 
-    fs->pathToType = strdup(dirPath);
-    array_new (fs->entityTypes, FILE_SERVICE_INITIAL_TYPE_COUNT);
+    // A recursive lock - fileServiceLoad() will recursively call fileServiceSave()
+    {
+        pthread_mutexattr_t attr;
+        pthread_mutexattr_init(&attr);
+        pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
 
+        pthread_mutex_init(&fs->lock, &attr);
+        pthread_mutexattr_destroy(&attr);
+    }
+
+    // Set the error handler - early
     fileServiceSetErrorHandler (fs, context, handler);
 
+    fs->sdb = NULL;
+    fs->sdbPath = NULL;
+
+    // Save currency and network
+    fs->currency = strdup (currency);
+    fs->network  = strdup (network);
+
+    // Locate the SQLITE Database
+    size_t sdbPathLength = strlen (basePath) + 1 + strlen(currency) + 1 + strlen(network) + 1 + strlen (FILE_SERVICE_SDB_FILENAME) + 1;
+    fs->sdbPath = malloc (sdbPathLength);
+    sprintf (fs->sdbPath, "%s/%s-%s-%s", basePath, currency, network, FILE_SERVICE_SDB_FILENAME);
+
+    pthread_mutex_lock (&fs->lock);
```

Comment:
Nothing wrong with it but curious, why do we lock in the "ctor"? Is a thread spawned off that could be accessing simultaneously?

Reply:
Fixed.